### DISCUSSION TITLE
Sending multiple parameters at Pipeline

### DISCUSSION THREAD

**Discussion by hasansoyalan** [laravel/framework#45302] [Ideas]
Hello,

Since pipeline accept single parameters, i would like to propose adding multiple parameters to Pipeline->send().  

```PHP
app(Pipeline::class)
    ->send($content, $user)
    ->through(FixMissingUsersPipe::class)
    ->then(function ($content, $user) { // Sending Multiple parameters
        // save to database etc...
    });
```

And usage is : 

```PHP
class FixMissingUsers
{
    public function handle(Content $content, User $user, $next)
    {
        // fix user
        return $next($content, $user);
    }
}
```

**Comment by henzeb** [comment]
Or you use a DTO. There is nothing wrong with using DTO's in this scenario. Could be as simple as a stdClass object

**Reply by shaedrich** [reply]
@hasansoyalan There's also a blog on how to implement this: https://www.luckymedia.dev/blog/laravel-pipelines-sending-multiple-parameters

**Comment by Jagdish-J-P** [comment]
@

> Hello,
> 
> Since pipeline accept single parameters, i would like to propose adding multiple parameters to Pipeline->send().
> 
> ```
> app(Pipeline::class)
>     ->send($content, $user)
>     ->through(FixMissingUsersPipe::class)
>     ->then(function ($content, $user) { // Sending Multiple parameters
>         // save to database etc...
>     });
> ```
> 
> And usage is :
> 
> ```
> class FixMissingUsers
> {
>     public function handle(Content $content, User $user, $next)
>     {
>         // fix user
>         return $next($content, $user);
>     }
> }
> ```

Try using it like:

```php
class FixMissingUsers
{
    public function handle(Content $content, $next, ...$parameters)
    {
        // fix user
        return $next($content, $parameters);
    }
}
```

Here you will receive array of multiple arguments in `$parameters` variable.

**Reply by henzeb** [reply]
Doesn't work. send does not allow you to add a second parameter.  
It does however work for strings. So if he would pass `FixMissingUsersPipe:1234` and then resolve the userId within the pipe. 

Another solution would be passing it to the constructor
````php
app(Pipeline::class)
    ->send($content)
    ->through(new FixMissingUsersPipe($user))
    ->then(function ($content) use ($user) { // Sending Multiple parameters
        // save to database etc...
    });
````

**Reply by Jagdish-J-P** [reply]
You can try using DTO or check this [link 1](https://laravelpipeline.com/basic-usage.html) here you can create transport like `UserRegistrationTransport` or [link 2](https://darkghosthunter.medium.com/laravel-when-to-use-pipelines-how-to-test-them-8993ed93e735) here you can create class with constructor like `SubscribeProcedure`.

**Reply by henzeb** [reply]
Thanks for the answer.... ==>>>> https://github.com/laravel/framework/discussions/45302#discussioncomment-4401622

**Comment by shaedrich** [comment]


**Reply by henzeb** [reply]
you are binding it to the closure that is executed at the end, not during.  and send still doesn't accept 2 parameters.

**Reply by shaedrich** [reply]
Ah, you're right 🤦🏻‍♂️

**Comment by msamgan** [comment]
```php

->through([
    new (QueryFilter::class)($q),
])

// Query filter class 

protected string $q;

public function __construct(
    $q
) {
    $this->q = $q;
}

```
this works and through this, you can pass any number of parameters.
